Gp/feat/aman arbitrary path

iparask opened the pull request:
This PR updates AxisManager `__getitem__` and `__setitem__` based on the slack discussion.

`__setitem__` raises an ValueError when the value is not another AxisManager and uses wrap to add it to the main object. I selected wrap because it seemed more complete compared to merge.

mhasself reviewed (changes requested):
Looks good -- a couple of requests though:

- `get` and `__contains__` have not been changed.  I think we should support the same treatment for keys provided there (but this is probably debatable).
- The tests don't test very much right now... please expand.
- Mention this behavior in the docs (see axisman.rst, add a few lines after paragraph of   "Note the boresight entry is marked with")

Thanks!

skhrg reviewed (commented):
I think mattew covered everything I had in mind.

I think the answer to this is "not enough that we care" or "not enough to measure" but is there any noticeable slowdown from accessing children in this way?

Expanding the test case so show that the values that we get and set are correct I think is important.

iparask commented:
> * `get` and `__contains__` have not been changed.  I think we should support the same treatment for keys provided there (but this is probably debatable).

Get is a combination of `__contains__` and `__getitem__` the way it is implemented. It's behavior change is a side effect of the rest of the changes.

> I think the answer to this is "not enough that we care" or "not enough to measure" but is there any noticeable slowdown from accessing children in this way?

Yes, this is measurable but how many levels down do you want to go?

skhrg commented:
> Yes, this is measurable but how many levels down do you want to go?

I think most people will never go being triple nested so if that slowdown is in the noise then I think we are good. If it somehow matters this seems like a case where we could cache the mapping from the full path to where they point to alleviate things (but I really dont think it will end up being an issue).

mhasself commented:
> > * `get` and `__contains__` have not been changed.  I think we should support the same treatment for keys provided there (but this is probably debatable).
> 
> Get is a combination of `__contains__` and `__getitem__` the way it is implemented. It's behavior change is a side effect of the rest of the changes.

Having changed `__getitem__`, we need to think about whether to change `.keys` (no?) and `__contains__` (yes?).

It is nice to be able to check if `k in aman` -- answering the question "will aman[k] raise a KeyError"?

mhasself commented:
> Is this what you mean? It is in the commit that addresses the comments. I'm trying to get them all in a single push.

Yes, that's what I mean.  Thanks.

mhasself commented on tests/test_core.py:
```diff
@@ -298,6 +298,44 @@ def test_300_restrict(self):
+            aman["child.dets.an_extra_layer"]
+
+        self.assertIn("child.dets", aman)
+        self.assertIn("child.dets2", aman) # I am not sure why this is true
```
This is true because axes are shared throughout an entire AxisManager structure.  When child2 was added to child, dets2 was made available to child and to aman.

mhasself commented on tests/test_core.py:
```diff
@@ -298,6 +298,44 @@ def test_300_restrict(self):
+        self.assertIn("child.dets2", aman) # I am not sure why this is true
+        self.assertNotIn("child.child2.someentry", aman)
+
+        aman["child"] = child2
```
This is a weird and dangerous case -- because AxisManagers have axes, and those axes need to be kept in sync, it's odd to set this directly.  One should "always use wrap when adding an AxisManager to an AxisManager".  So I might remove this test, since it seems to legitimize unsafe usage.

iparask replied:
I will raise a ValueError when that is happening. Is there a valid case? If you see the diff in `__setitem__` it is not that different apart from going through the path.

mhasself reviewed (changes requested):
Thanks.

I think what's missing in the test is just a simple data field, such as a shape=(samps,) array, in child or child2.  Check that accessing, and setting, that field using 'child.child2.target' works properly.

iparask commented:
I updated the tests and added a `_fields` tod to the test. I'm really confused as to why `__setattr__` and `__setitem__` have different functionalities. It seems like a bug.

I also timed them, using timeit and 1000 iterations and these are the results:
```
(simonsobs) iparask@simons1:~/git/sotodlib$ python tests/simple_example.py
Time to access aman['child.child2.tod'] from this branch: 0.0009413734078407288

(simonsobs) iparask@simons1:~/git/sotodlib$ python tests/simple_example.py
Time to access aman.child.child2.tod from sotodlib master branch: 0.0019882768392562866
```

I consider these to be the same time.

iparask commented on docs/axisman.rst:
```diff
@@ -144,6 +144,17 @@ The output of the ``wrap`` cal should be::
 Note the boresight entry is marked with a ``*``, indicating that it's
 an AxisManager rather than a numpy array.
 
+To access data in an AxisManager, use a path-like syntax where 
```
Fixed. I added a simple data access sentence and an advances access sentence

skhrg replied:
This is super clear thanks!

iparask commented:
@mhasself can you take another look?

mhasself commented on docs/axisman.rst:
```diff
@@ -144,6 +144,26 @@ The output of the ``wrap`` cal should be::
 Note the boresight entry is marked with a ``*``, indicating that it's
 an AxisManager rather than a numpy array.
 
+Data access under an AxisManager is done based on field names. For example:
```
In your example, you construct a new aman.  But in the flow of this documentation, an AxisManager already exists, which has been built up demonstrating the features and usage.  The existing one has enough stuff to demonstrate all the features you need to demonstrate (e.g. `dset.boresight.az` vs. `dset['boresight.az']`).  So please use that!  (And that should flow into the "To slice this object..." section, which again refers to the `dset` variable.)

Also this line for some reason renders in bold, in the docs... maybe add a blank line after?  Not sure how you got away without a double `::`...

iparask replied:
Done

mhasself reviewed (changes requested):
Thanks -- a few more comments.

iparask commented on tests/test_core.py:
```diff
@@ -298,6 +299,56 @@ def test_300_restrict(self):
+
+        self.assertIn("child.dets", aman)
+        self.assertIn("child.dets2", aman) # I am not sure why this is true
+        self.assertNotIn("child.child2.someentry", aman)
```
Done

iparask commented on sotodlib/core/axisman.py:
```diff
@@ -349,29 +349,60 @@ def move(self, name, new_name):
+        if isinstance(val, AxisManager) and isinstance(tmp_item, AxisManager):
+            raise ValueError("Cannot assign AxisManager to AxisManager. Please use wrap method.")
+
+        tmp_item.__setattr__(val_key, val)
```
done, I think

iparask commented on tests/test_core.py:
```diff
@@ -298,6 +299,56 @@ def test_300_restrict(self):
         self.assertNotEqual(aman.a1[0, 0, 0, 1], 0.)
 
     # wrap of AxisManager, merge.
+    def test_get_set(self):
```
Done

iparask commented on tests/test_core.py:
```diff
@@ -3,6 +3,7 @@
 import os
 import shutil
 
+from networkx import selfloop_edges
```
Removed

mhasself reviewed (approved):
Looks good, thanks for the iterations.

Obviously we need to get CI working again but otherwise this is good to go.

Remember to squash on merge.

skhrg reviewed (approved):
Looks great to me!
